lib/common/zstd_internal.h

```diff
@@ -352,7 +352,7 @@ typedef enum {
 *  Private declarations
 *********************************************/
 typedef struct seqDef_s {
-    U32 offset;         /* Offset code of the sequence */
+    U32 offset;         /* offset == rawOffset + ZSTD_REP_NUM, or equivalently, offCode + 1 */
```

Comment:
I think it's `offCode + 2`

Reply:
Are the values contained within the `rep` array not considered to be raw offsets? The terminology used is a bit confusing, but I've found that I need to add 3 == ZSTD_REP_NUM to the value within `rep` to obtain the correct value to give `seqDef::offset`, and that `ZSTD_updateRep()` expects an `offCode` which appears to be `offset - 1`?

`seqDef::offset` can have value == 1 which should correspond to `offCode == 0`, right?

Reply:
That's one of the messiest confusing part of the code,
and it is probably worth a PR just dedicated to fix it.

Initially, `offset` was really that, an `offset`. See value `16`, it means `offset == 16`.
Quite logically, the name `offset` was adopted to represent this value everywhere it's needed and stored.

There was an exception though, as `0` can't be a valid offset. 
So it was refurbished to mean `repCode`. 
Back then, there was only a single repcode value available. 
So OK, `offset` means `offset` when the value `> 0`, and there is a special case when `offset==0`.
Still fair enough.

Later on, much later in the life of pre-v1.0 `zstd`, Przemyslaw added the ability to manage 3 repcodes.
This was beneficial for the high compression mode, resulting in small but measurable benefits.
It complexifies repcode history update, but Przemyslaw also worked that out pretty well, so speed was barely affected.

But hey, the associated variable was already called `offset` and used in multiple place over the code base,
so the name was left as is. 
Except that now, it's no longer an offset. It's effectively a sum type, that means either a `repcode`, or an offset.

At the time, it felt like a minor, localized, source of confusion and complexity, and we could live with it.
Sure. But as the code ages, and its nb of contributors increases, this little quirk feels less and less acceptable.

> Are the values contained within the rep array not considered to be raw offsets? 

I presume "rep array" means the 3-elts repcode history array ?
I don't remember precisely, but I would expect them to be raw offsets, since the `rep` array shouldn't contain `repcodes`.

However, I'm surprised you bring that up, as this comment is written in the context of `seqDef`, which corresponds to the Sequence Store. So I don't see a direct relation with the repcode history.

If I remember correctly, in the `seqStore`, the `offset` values `[0, 1, 2]` actually correspond to the 3 repcodes. Then starting with code `3`, the "real" offset range starts, with this first valid offset being 1. This makes a difference of 2, which corresponds to `ZSTD_REP_MOVE`.

Reply:
I appreciate the explanation on this topic! Interesting stuff here...

> If I remember correctly, in the seqStore, the offset values [0, 1, 2] actually correspond to the 3 repcodes. Then starting with code 3, the "real" offset range starts, with this first valid offset being 1. This makes a difference of 2, which corresponds to ZSTD_REP_MOVE.

I guess what I'm saying is that what I've observed is that in the `seqStore`, the `offset` values `[1, 2, 3]` actually correspond to the 3 repcodes, rather than `[0, 1, 2]` as you mentioned above. Thus, the "real" offset range starts with code `4`.

If we look at `zstd_fast.c` for an example of converting a true raw offset into a `seqStore->offset`, we start by adding `ZSTD_REP_MOVE` to the raw offset value, getting an `offCode`, as you mentioned above:
https://github.com/facebook/zstd/blob/56421f34e4ab5c17cbac0832bd12413e4f27ecf9/lib/compress/zstd_fast.c#L143

And then store `offCode` as such, with `ZSTD_storeSeq()`:
https://github.com/facebook/zstd/blob/56421f34e4ab5c17cbac0832bd12413e4f27ecf9/lib/compress/zstd_fast.c#L152

But if we take a look at the actual implementation of `ZSTD_storeSeq()`, there's something funny here:

In `ZSTD_storeSeq()`, we increment what we actually store:
https://github.com/facebook/zstd/blob/56421f34e4ab5c17cbac0832bd12413e4f27ecf9/lib/compress/zstd_compress_internal.h#L619

So, we actually end up adding `3 == ZSTD_REP_NUM` to the raw offset to obtain `seqStore->offset`, so `rawOffset + 3 == offCode + 1`, right? It's not actually clear to me why we increment it. Maybe it is a left-over from back before repcodes existed as you mentioned, and never actually got fully fixed to support `offset == 0` and so we still have that `MIN(seqStore->offset) == 1` for some reason. Because the scheme that you described in your comment makes more sense than what currently seems to exist.

Reply:
The `offCode` gets broken up into two pieces. The code `nbBits = ZS_highbit32(offCode)` and the `nbBits` extra bits. `ZS_highbit32(0) = undefined & ZS_highbit32(1) = 0`. So we need to make sure that `offCode` is never zero, so we add 1.

So on compression side, we use offset == 1, 2, 3 for repcodes. On decompression side it is offsets 0, 1, 2 for repcodes. They correspond to the same thing, since we've shifted up by one on the compression side.

Reply:
Ah that makes a lot of sense, thanks for the clarification.